### DISCUSSION TITLE
What defines the order of sessions in mini.starter?

### DISCUSSION THREAD

**Discussion by dathinaios** [nvim-mini/mini.nvim#1354] [Q&A] [answered]
Just a question: I am using `mini.starter` to simply display my saved sessions (normal sessions not `mini.session`). The last session that I have saved appears first but I would like to be able to reorder. How can I do it?

**Comment by echasnovski** [comment]
@dathinaios, could you, please, clarify what setup you are using?

The default items contain "Sessions" section which [displays only sessions detected with 'mini.sessions'](https://github.com/echasnovski/mini.nvim/blob/6714e738555b1cfbef646a23af29aa113f91f4fe/doc/mini-starter.txt#L340). It's default order is indeed from most to least recent, but can be [changed to alphabetical](https://github.com/echasnovski/mini.nvim/blob/6714e738555b1cfbef646a23af29aa113f91f4fe/doc/mini-starter.txt#L355). Other than that you'd have to write your own section with any order you'd want.

**Reply by dathinaios** [reply]
Thank you for the answer. I have quite a custom configuration that you helped me with in the past:

```lua
local header_art =
[[
 ╭╮╭┬─╮╭─╮┬  ┬┬╭┬╮
 │││├┤ │ │╰┐┌╯││││
 ╯╰╯╰─╯╰─╯ ╰╯ ┴┴ ┴
]]

-- using the mini plugins
require('mini.sessions').setup({
  -- Whether to read latest session if Neovim opened without file arguments
  autoread = false,
  -- Whether to write current session before quitting Neovim
  autowrite = false,
  -- Directory where global sessions are stored (use `''` to disable)
  directory =  '~/.vim/sessions', --<"session" subdir of user data directory from |stdpath()|>,
  -- File for local session (use `''` to disable)
  file = '' -- 'Session.vim',
})

local starter = require('mini.starter')
starter.setup({
  -- to disable letter navigation so I can map j,k
  -- query_updaters = '',
  -- evaluate_single = true,
  items = {
    starter.sections.sessions(77, true),
    starter.sections.builtin_actions(),
  },
  content_hooks = {
    function(content)
      local blank_content_line = { { type = 'empty', string = '' } }
      local section_coords = starter.content_coords(content, 'section')
      -- Insert backwards to not affect coordinates
      for i = #section_coords, 1, -1 do
        table.insert(content, section_coords[i].line + 1, blank_content_line)
      end
      return content
    end,
    starter.gen_hook.adding_bullet("» "),
    starter.gen_hook.aligning('center', 'center'),
  },
  header = header_art,
  footer = '',
})

-- fix from AI
vim.api.nvim_create_augroup('MiniStarterJK', { clear = true })

vim.api.nvim_create_autocmd('User', {
  pattern = 'MiniStarterOpened',
  callback = function()
    if vim.bo.filetype == 'ministarter' and vim.api.nvim_buf_get_name(0):match('Starter') then
      vim.api.nvim_buf_set_keymap(0, 'n', 'j', "<Cmd>lua MiniStarter.update_current_item('next')<CR>", { noremap = true, silent = true })
      vim.api.nvim_buf_set_keymap(0, 'n', 'k', "<Cmd>lua MiniStarter.update_current_item('prev')<CR>", { noremap = true, silent = true })
    end
  end
})
```

(Just noticed that I am actually using `mini.sessions`.. 😬)

**Reply by echasnovski** [reply, marked as answer]
> (Just noticed that I am actually using `mini.sessions`.. 😬)

Yeah, that appears to be the case. Then this is indeed relevant:

> The default items contain "Sessions" section which [displays only sessions detected with 'mini.sessions'](https://github.com/echasnovski/mini.nvim/blob/6714e738555b1cfbef646a23af29aa113f91f4fe/doc/mini-starter.txt#L340). It's default order is indeed from most to least recent, but can be [changed to alphabetical](https://github.com/echasnovski/mini.nvim/blob/6714e738555b1cfbef646a23af29aa113f91f4fe/doc/mini-starter.txt#L355). Other than that you'd have to write your own section with any order you'd want.

If you really need custom session order, then you can use a callable `MiniStarter.config.items` entry which reads [`MiniSessions.detected`](https://github.com/echasnovski/mini.nvim/blob/6714e738555b1cfbef646a23af29aa113f91f4fe/doc/mini-sessions.txt#L108) and converts it to array of ['mini.starter' items](https://github.com/echasnovski/mini.nvim/blob/6714e738555b1cfbef646a23af29aa113f91f4fe/doc/mini-starter.txt#L13-L18) in the order you want. The [relevant part of `MiniStarter.sections.sessions`](https://github.com/echasnovski/mini.nvim/blob/6714e738555b1cfbef646a23af29aa113f91f4fe/lua/mini/starter.lua#L454-L489) should be a good starting point: simplify to use hyperparameters and sorting as you'd like.

**Reply by dathinaios** [reply]
Thank you for the answer! I thought it might be a file with the items somewhere like `arrow.nvim` uses that I can invoke and edit. Maybe a useful feature? But I will give your suggestion a try. Thanks again! 🙂

**Reply by echasnovski** [reply]
I think having the Starter screen be configured inside `require('mini.starter').setup()` makes sense. Not least because items can be functions to be executed right before showing Starter buffer.

If you mean sessions, then they are detected [based on the config](https://github.com/echasnovski/mini.nvim/blob/6714e738555b1cfbef646a23af29aa113f91f4fe/doc/mini-sessions.txt#L15-L18), so having them as a list in a single file is not really possible. Setting proper [directory for global sessions](https://github.com/echasnovski/mini.nvim/blob/6714e738555b1cfbef646a23af29aa113f91f4fe/doc/mini-sessions.txt#L86) is enough and no extra file is needed.

**Reply by dathinaios** [reply]
Hello 🙂

The j, k bindings from above stopped working for me, possibly after updating to 0.11 though not sure if that is what broke it or a `Lazy` update. Did anything change that could cause the above to stop working maybe?

Thanks!

**Reply by echasnovski** [reply]
> Did anything change that could cause the above to stop working maybe?

If you're talking about [this code](https://github.com/echasnovski/mini.nvim/discussions/1354#discussioncomment-11339061), then removing the part of `and vim.api.nvim_buf_get_name(0):match('Starter')` should help. It was not really useful in the first place, but after the (not so recent) [update](https://github.com/echasnovski/mini.nvim/commit/7f2cc51e18f77d6154bedfb324146061bc38a472) this will always return `false` because the naming structure has changed.

**Reply by dathinaios** [reply]
That was it! Thank you so much 😀 I have customised the starter so much at this point.. 😅 I am really enjoying the experience. Again, thank you for the speedy reply!
